File: gred/src/texture_viewer.lua
TextureViewer = class()

function TextureViewer:Create(texture_names, selected, path, prev_key, next_key)
  self = TextureViewer:New()
  self.names = texture_names
  self.selected = Clamp(math.floor(selected), 1, #texture_names)
  self.path = path
  self.tex = nil
  self.prev_key = prev_key
  self.next_key = next_key
  self:_reloadTexture()
  return self
end

function TextureViewer:update()
  if IsKeyPressed(self.prev_key) then self:_prevTexture() end
  if IsKeyPressed(self.next_key) then self:_nextTexture() end
end

function TextureViewer:draw(x, y, width, height)
  if self.tex ~= nil then
    local src = Rectangle()
    src.x = 0
    src.y = 0
    src.width = self.tex.width
    src.height = self.tex.height
    local dst = Rectangle()
    dst.x = x
    dst.y = y
    dst.width = width
    dst.height = height
    DrawTexturePro(self.tex, src, dst, Vector2(), 0, GetColor(0xFFFFFFFF))
  end
end

function TextureViewer:texture()
  return self.tex
end

function TextureViewer:textureName()
  return self.names[self.selected]
end

function TextureViewer:_nextTexture()
  self.selected = self.selected + 1
  if self.selected > #self.names then self.selected = 1 end
  self:_reloadTexture()
end

function TextureViewer:_prevTexture()
  self.selected = self.selected - 1
  if self.selected < 1 then self.selected = #self.names end
  self:_reloadTexture()
end

function TextureViewer:_reloadTexture()
  self.tex = LoadTexture(self.path .. self.names[self.selected])
end

File: src/texture_viewer.cpp
#include "texture_viewer.h"
#include <algorithm>

TextureViewer::TextureViewer(const std::vector<std::string>& names_, int selected_,
                             const std::string& path_, int pk, int nk)
    : names(names_), path(path_), prev_key(pk), next_key(nk)
{
    selected = std::min(std::max(selected_, 1), (int)names.size());
    reload_texture();
}

void TextureViewer::update() {
    if (App::key_hit[prev_key]) prev_texture();
    if (App::key_hit[next_key]) next_texture();
}

void TextureViewer::draw(int x, int y, int w, int h) const {
    if (tex) App::draw_texture_ex(tex, x, y, w, h, COLOR_WHITE);
}

Texture* TextureViewer::texture() const { return tex; }

const std::string& TextureViewer::texture_name() const {
    static std::string empty;
    if (names.empty()) return empty;
    return names[selected - 1];
}

void TextureViewer::next_texture() {
    ++selected;
    if (selected > (int)names.size()) selected = 1;
    reload_texture();
}

void TextureViewer::prev_texture() {
    --selected;
    if (selected < 1) selected = (int)names.size();
    reload_texture();
}

void TextureViewer::reload_texture() {
    if (names.empty()) { tex = nullptr; return; }
    std::string full = path + names[selected - 1];
    tex = CacheTexture(full.c_str(), FILTER_LINEAR);
}
